promsonnet/legacy.libsonnet

```
@@ -0,0 +1,56 @@
+{
+  // In this file, the word 'update' refers to retrofitting maps and _order
+  // lists to data structures that lack them.
+
+  v1: {
+    local v1 = self,
+    ruleGroupSet: {
+      update():: {
+        local groups = super.groups,
+        local this = self,
+        local groupsAsMap(group, acc) = acc { [group.name]+: group + v1.ruleGroup.update() },
+        local groupsAsList(group, acc) = acc + [group.name],
+        groups_map:: std.foldr(groupsAsMap, groups, {}),
+        groups_order:: std.foldr(groupsAsList, groups, []),
+        groups: [this.groups_map[group] for group in this.groups_order],
+      },
+    },
+
+    patchRule(group, rule, patch):: {
+      groups: std.map(function(g)
+                        if g.name == group
+                        then g {
+                          rules: std.map(function(r)
+                                           if ('alert' in r && r.alert == rule) || ('record' in r && r.record == rule)
+                                           then r + patch
+                                           else r,
+                                         g.rules),
+                        }
+                        else g,
+                      super.groups),
+    },
```

Add a few more newlines 
```suggestion
    patchRule(group, rule, patch):: {
      groups: std.map(
        function(g)
          if g.name == group
          then g {
            rules: std.map(
              function(r)
                if ('alert' in r && r.alert == rule) ||
                   ('record' in r && r.record == rule)
                then r + patch
                else r,
              g.rules
            ),
          }
          else g,
        super.groups
      ),
    },
```